### PR TITLE
Exchange

### PR DESCRIPTION
**Pull request by jenyafomin**

### CODE REVIEW

File: admin/complex-mechanics/pyramid.md [hunk 1]
```diff
@@ -0,0 +1,15 @@
+---
+sidebar_position: 2
+---
+
+# Pyramid
+
+Pyramid is a gambling mechanic based on Ponzi scheme. 
+
+## Deployment
```
Comment (TrejGun) [inline admin/complex-mechanics/pyramid.md:9]: Staking and pyramid is not mean to be deployed on B2C version, please this section from two pages

File: admin/exchange/_category_.json [hunk 2]
```diff
@@ -0,0 +1,8 @@
+{
+  "label": "Exchange",
+  "position": 5,
+  "link": {
+    "type": "generated-index",
+    "description": "This section explain the main concept for exchanging items between User and  different mechanics like Exchange(marketplace), Staking, etc."
```
Comment (TrejGun) [inline admin/exchange/_category_.json:6]: between User and the System by different mechanics

File: admin/exchange/overview.md [hunk 3]
```diff
@@ -0,0 +1,24 @@
+---
+sidebar_position: 1
+---
+
+# Overview
+
+Exchange contract allow users to exchange their items for other items or assets, depending on the specific mechanics used in the contract. The table below provides an overview of several common mechanics that use exchange functionality:
```
Comment (TrejGun) [inline admin/exchange/overview.md:7]: allowS?

Comment (TrejGun) [reply admin/exchange/overview.md:7]: their assets for other assets

Comment (TrejGun) [reply admin/exchange/overview.md:7]: uses?

File: admin/simple-mechanics/wrapper.md [hunk 4]
```diff
@@ -0,0 +1,9 @@
+---
+sidebar_position: 8
+---
+
+# Wrapper
+
+The main idea of the **Wrapper** mechanic is that it can **wrap an unlimited amount of [assets](/admin/miscellaneous/asset/)** into a single token. The **Wrap token** represents all of the [assets](/admin/miscellaneous/asset/) that were wrapped, and can be transferred to anyone in a single transfer. When a user wants to unwrap the Wrap token, they can call the "**unpack**" function and pass the **token ID** of the Wrap token. This will transfer all of the wrapped tokens back to the owner's address, and the *Wrap token* will be **burned**.
```
Comment (TrejGun) [inline admin/simple-mechanics/wrapper.md:7]: remove "an unlimited amount of" it is technically limited by transaction gas limit

File: api/exchange.md [hunk 5]
```diff
@@ -0,0 +1,59 @@
+# Exchange
+
+Exchange is the main contract that serves the mechanics around the tokens.
```
Comment (TrejGun) [inline api/exchange.md:3]: Exchange is the one of core contracts

File: api/exchange.md [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+This contract enables the exchange of tokens in a flexible manner, with various exchange mechanics available such as **[Claim](/admin/simple-mechanics/claim/)**, **[Grade](/admin/simple-mechanics/grade/)**, **Purchase**, etc. This functionality allows users to exchange one or multiple assets. However, to prevent unauthorized transactions, a trusted source (Server) must sign the transaction before it can be executed on the blockchain. Once the signature and additional information from the Server response are obtained, they can be combined with the Asset information to execute the exchange on the blockchain.
+
+#### Workflow of Purchase mechanics.
+![](/img/exchange_diagram.jpeg)
+
+
+### Signing a Purchase Transaction:
+
+To sign any transaction, we use the standard [EIP-712](https://eips.ethereum.org/EIPS/eip-712). This combines all parameters passed to the contract function into one string and hashes them. When executing the transaction, the address will be recovered from signature and contract will check this address for the appropriate minting rights. If the address has the necessary rights, the transaction is processed.
```
Comment (TrejGun) [inline api/exchange.md:15]: to the contract function and hashes them

Comment (TrejGun) [reply api/exchange.md:15]: and contract will check this address for the appropriate permissions

Comment (TrejGun) [reply api/exchange.md:15]: the address has the necessary permissions

File: api/exchange.md [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+To sign a transaction via the Server endpoint, several arguments must be provided, including the **account address**, the **referral address** (if not applicable, pass '0x'), and the **templateId**.
```
Comment (TrejGun) [inline api/exchange.md:17]: (if not applicable, pass nothing or '0x'),

File: api/exchange.md [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
+> Different mechanics mey require different input. For Excample **Craft** require only **craftId**.
```
Comment (TrejGun) [inline api/exchange.md:19]: may

>>>>
all mechanics requires account and optional referrer plus specific id of mechanic

File: api/exchange.md [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+### Overview of Exchange Functionality:
+
+This contract enables the exchange of tokens in a flexible manner, with various exchange mechanics available such as **[Claim](/admin/simple-mechanics/claim/)**, **[Grade](/admin/simple-mechanics/grade/)**, **Purchase**, etc. This functionality allows users to exchange one or multiple assets. However, to prevent unauthorized transactions, a trusted source (Server) must sign the transaction before it can be executed on the blockchain. Once the signature and additional information from the Server response are obtained, they can be combined with the Asset information to execute the exchange on the blockchain.
+
+#### Workflow of Purchase mechanics.
+![](/img/exchange_diagram.jpeg)
+
+
+### Signing a Purchase Transaction:
```
Comment (TrejGun) [inline api/exchange.md:13]: add example of request with templateId

File: api/exchange.md [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+
+**First** of all need to check - ***ExchangeContract* have allowance** for all tokens in **price**. If not we have to **[approve](/market/miscellaneous/approve/) tokens**. 
+
+**Second** we have call function that will perform the purchase. 
+In order to create an instance of the ***ExchangeContract*** we need **address** and **ABI** of ***ExchangeContract***.
+
+All contracts in our ecosystem use the same logic for exchange and have almost the same arguments for executing:
+- **params**:
+    - **nonce**         : Unique request ID,
+    - **externalId**    : This is the **ID in the database** that applies to the specific mechanic. In the case of **Purchase**, the externalId would be equal to the **templateId** of the token being purchased. For **Craft**, the externalId would be equal to the **craftId** of the crafting recipe being used.
```
Comment (TrejGun) [inline api/exchange.md:42]: of the template (LINK to 721 template) being purchased

File: api/exchange.md [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+
+**Second** we have call function that will perform the purchase. 
+In order to create an instance of the ***ExchangeContract*** we need **address** and **ABI** of ***ExchangeContract***.
+
+All contracts in our ecosystem use the same logic for exchange and have almost the same arguments for executing:
+- **params**:
+    - **nonce**         : Unique request ID,
+    - **externalId**    : This is the **ID in the database** that applies to the specific mechanic. In the case of **Purchase**, the externalId would be equal to the **templateId** of the token being purchased. For **Craft**, the externalId would be equal to the **craftId** of the crafting recipe being used. 
+    - **expiresAt**     : This parameter describes how long signature would be valid _(If equal to 0, don't have any time limitations)_,
+    - **referrer**      : This is the **address** of the person who invited the account to the platform. If a referral program is not applicable, the value '0x' should be passed as the referrer. The referrer is used to track referrals and reward the person who made the referral.
```
Comment (TrejGun) [inline api/exchange.md:44]: l program is not installed

File: api/exchange.md [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+- **item**: [Asset](/admin/miscellaneous/asset/) that user will receive,
+- **price**: an array of [Assets](/admin/miscellaneous/asset/) that user must transfer,
```
Comment (TrejGun) [inline api/exchange.md:46]: remove "an array of "

File: api/exchange.md [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+- **sign**: This is a hash of all the arguments listed above but is was generated on the server and must contain the same parameters as listed above, with the exception of signature itself. The signature ensures the authenticity of the transaction and prevents tampering or modification of the transaction data during transmission.
+
+
+#### Some notes:
+
+> If **Native tokens** are in the **"price"**, the transaction must be executed with the same value as in **price.amount**.
```
Comment (TrejGun) [inline api/exchange.md:52]: as a sum of all **price.amount**.

File: api/exchange.md [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+
+> If the Exchange functionality (Purchase, Grade, etc.)  **accepts an array** of **"item"** or **"price"**, **unlimited [assets](/admin/miscellaneous/asset/) can be passed** for transfer or received as long as there is enough gas in the transaction.
```
Comment (TrejGun) [inline api/exchange.md:54]: remove unlimited

File: admin/complex-mechanics/pyramid.md (hunk 1 above)
Comment (jenyafomin) [reply admin/complex-mechanics/pyramid.md:9]: For this need a little bit more information

**Pull request merged**